fs.go

```diff
@@ -17,96 +17,521 @@
+		diskInfo:           diskInfo,
+		minFreeDisk:        5, // Minimum 5% disk should be free.
+		listObjectMap:      make(map[listParams][]*treeWalker),
+		listObjectMapMutex: &sync.Mutex{},
+		rwLock:             &sync.RWMutex{},
+	}
+	return fs, nil
+}
 
-	// Initialize list object map.
-	fs.listObjectMap = make(map[listObjectParams][]*treeWalker)
-	fs.listObjectMapMutex = &sync.Mutex{}
+// checkDiskFree verifies if disk path has sufficient minium free disk space.
+func checkDiskFree(diskPath string, minFreeDisk int64) (err error) {
+	di, err := disk.GetInfo(diskPath)
+	if err != nil {
+		return err
+	}
 
-	// Initialize list multipart map.
-	fs.listMultipartObjectMap = make(map[listMultipartObjectParams][]<-chan multipartObjectInfo)
-	fs.listMultipartObjectMapMutex = &sync.Mutex{}
+	// Remove 5% from total space for cumulative disk
+	// space used for journalling, inodes etc.
+	availableDiskSpace := (float64(di.Free) / (float64(di.Total) - (0.05 * float64(di.Total)))) * 100
+	if int64(availableDiskSpace) <= minFreeDisk {
+		return errDiskPathFull
+	}
 
-	// Return here.
-	return fs, nil
+	// Success.
+	return nil
 }
 
-func (fs Filesystem) checkBucketArg(bucket string) (string, error) {
-	if !IsValidBucketName(bucket) {
-		return "", BucketNameInvalid{Bucket: bucket}
+// checkVolumeArg - will convert incoming volume names to
+// corresponding valid volume names on the backend in a platform
+// compatible way for all operating systems. If volume is not found
+// an error is generated.
+func (s fsStorage) checkVolumeArg(volume string) (string, error) {
+	if !isValidVolname(volume) {
+		return "", errInvalidArgument
+	}
+	volumeDir := filepath.Join(s.diskPath, volume)
+	_, err := os.Stat(volumeDir)
+	if err == nil {
+		return volumeDir, nil
 	}
-	bucket = getActualBucketname(fs.diskPath, bucket)
-	if status, e := isDirExist(filepath.Join(fs.diskPath, bucket)); !status {
-		if e == nil {
-			return "", BucketNotFound{Bucket: bucket}
-		} else if os.IsNotExist(e) {
-			return "", BucketNotFound{Bucket: bucket}
-		} else {
-			return "", e
+	if os.IsNotExist(err) {
+		var volumes []os.FileInfo
+		volumes, err = ioutil.ReadDir(s.diskPath)
+		if err != nil {
+			return volumeDir, errVolumeNotFound
 		}
+		for _, vol := range volumes {
+			if vol.IsDir() {
+				// Verify if lowercase version of the volume
+				// is equal to the incoming volume, then use the proper name.
+				if strings.ToLower(vol.Name()) == volume {
+					volumeDir = filepath.Join(s.diskPath, vol.Name())
+					return volumeDir, nil
+				}
+			}
+		}
+		return volumeDir, errVolumeNotFound
+	} else if os.IsPermission(err) {
+		return volumeDir, errVolumeAccessDenied
 	}
-	return bucket, nil
+	return volumeDir, err
 }
 
-func checkDiskFree(diskPath string, minFreeDisk int64) error {
-	di, e := disk.GetInfo(diskPath)
-	if e != nil {
+// Make a volume entry.
+func (s fsStorage) MakeVol(volume string) (err error) {
+	volumeDir, err := s.checkVolumeArg(volume)
```

Comment:
its better to have func `getVolumeDir` for given volume name.  it looks like `s.CheckVolumeArg()` returns `(bool, error)`